### ISSUE TITLE
Can't configure lsp servers with file/directory local variables

### ISSUE THREAD

**Issue by duog** [emacs-lsp/lsp-mode#405]
Hi,

I've been trying to configure the haskell lsp server with directory local variables and have run into problems.

It seems that directory local variables are [after major-mode hooks are run](https://stackoverflow.com/questions/19280851/how-to-keep-dir-local-variables-when-switching-major-modes). This means that if i use the natural
```
(add-hook 'haskell-mode 'lsp-haskell-enable)
```
then the server will be started before the path to its executable is configured by my .dir-locals.el.

There are [workarounds](https://stackoverflow.com/questions/5147060/how-can-i-access-directory-local-variables-in-my-major-mode-hooks), however I find these unsatisfying since everyone (who wants to use .dir-locals.el like this) has to remember to do it every time they enable an lsp-server.

Flycheck seems to solve this same problem by -- instead of running a check -- [setting a deferred flag on initialisation](https://github.com/flycheck/flycheck/blob/f6f091949cdaf89f228af46f1562c1438e2b1890/flycheck.el#L2404) and [checking that flag](https://github.com/flycheck/flycheck/blob/f6f091949cdaf89f228af46f1562c1438e2b1890/flycheck.el#L2354) on `post-command-hook` and `window-configuration-change-hook`. Note that I have not been able to find anything to say that they do this to support local variables.

Would you consider using a similar strategy in the implementation of `lsp--enable-stdio-client`?

**Comment by yyoncho** [comment]
I might be missing something, but wouldn't adding ...

``` emacs-lisp
(run-hooks (intern (concat (symbol-name major-mode) "-local-vars-hook")))
```
... to the *-enable function do the job as per the link you have provided?

**Comment by yyoncho** [comment]
Will be solved by #544. It will allow the following configuration:

```
(add-hook 'hack-local-variables-hook (lambda () (when (derived-mode-p 'prog-mode) (lsp))))
```

**Comment by yyoncho** [comment]
Fixed with in latest lsp-mode, sample configuration:

```
(add-hook 'hack-local-variables-hook
          (lambda () (when (derived-mode-p 'XXX-mode) (lsp))))
```

**Comment by S-D-** [comment]
Hi, this doesn't work for me for some reason...
In my init.el I have:
```
(add-hook 'hack-local-variables-hook
          (lambda () (when (derived-mode-p 'prog-mode) (print ccls-initialization-options) (lsp))))
```
and in .dir-locals.el:
```
((prog-mode
  (ccls-initialization-options . '(:compilationDatabaseDirectory "build"))))
```
But after I open a source file in the directory in `*ccls::sstderr*` I see:
```
14:51:39 ccls           initialize.cc:294 I initializationOptions: {"compilationDatabaseCommand":"","compilationDatabaseDirectory":"", ...
```
While in `ccls-initialization-options` variable help i see the value set correctly.

The only way to make `ccls-initialization-options` set correctly is to do
```
(add-hook 'prog-mode-hook #'lsp)
(setq ccls-initialization-options '(:compilationDatabaseDirectory "build")
```
in my init.el, but this configures `ccls-initialization-options` globaly...

**Comment by yyoncho** [comment]
@S-D- what does `(print ccls-initialization-options)` outputs?

**Comment by S-D-** [comment]
@yyoncho it outputs `(quote (:compilationDatabaseDirectory "build")` as expected

**Comment by yyoncho** [comment]
@S-D- can you set lsp-print-io to t and then inspect the "initialize" message (it will be located in "\*lsp-log: ccls " buffer

**Comment by S-D-** [comment]
Ah, I found a mistake...
Here is the initialization params from "lsp-log: ccls" buffer:
```
[Trace - 05:13:48 PM]
Sending request 'initialize - (1)'.

Params: 
{
  "processId": 13086,
  ...
  "initializationOptions": [
    "quote",
    {
      "compilationDatabaseDirectory": "build"
    }
  ]
}
```
Looks like I shouldn't use quote in .dir-locals.el. After changing .dir-local.el to
```
((prog-mode
  (ccls-initialization-options . (:compilationDatabaseDirectory "build"))))
```
everything works as expected.
Thanks for your help, @yyoncho!

**Comment by yyoncho** [comment]
@S-D-  You welcome (as a side note there is function add-dir-local-variable )

**Comment by hg-zt** [comment]
@yyoncho Hi, I'm using the corrected configuration, but it still not work.

```
  (add-hook 'hack-local-variables-hook
            (lambda () (when (derived-mode-p 'prog-mode) (print ccls-initialization-options) (lsp))))
```

```
((prog-mode
  (ccls-initialization-options . (:compilationDatabaseDirectory "build"))))
```

I've checked `*lsp-log: ccls` buffer and I could not even see this option added into the config.

This is what I see in message buffer, it seems that the `ccls` process is started before the value is overwritten by `.dir-locals.el` ?
```
Configuring package lsp-ui...done
LSP :: Connected to [ccls:14171 status:starting].

(:compilationDatabaseDirectory "build")

LSP :: Connected to [ccls:14171 status:starting].

```

I've also tried to change `ccls-args`, which should set the launch command with args, but still no luck: I set it to "--log-file=/tmp/ccls.log" in global conf and "--log-file=/tmp/ccls2.log" in local conf, and log is always written to "/tmp/ccls.log".

**Comment by yyoncho** [comment]
@hg-zt I think that you have another hook that calls `lsp`. You may find out what is the hook by doing `M-x debug-on-entry RET lsp RET`

**Comment by hg-zt** [comment]
@yyoncho 

thanks for your fast reply.

yeah it breaks twice on lsp()

and the first backtrace is 

```
Debugger entered--entering a function:
* lsp()
  (lambda nil (if (memql (function company-lsp) company-backends) (with-no-warnings company-backends) (setq company-backends (cons (function company-lsp) company-backends))) (require (quote ccls)) (remhash (quote clangd) lsp-clients) (lsp))()
  run-hooks(change-major-mode-after-body-hook prog-mode-hook c-mode-common-hook c++-mode-hook)
  apply(run-hooks (change-major-mode-after-body-hook prog-mode-hook c-mode-common-hook c++-mode-hook))
  run-mode-hooks(c++-mode-hook)
  c++-mode()
  set-auto-mode-0(c++-mode nil)
  set-auto-mode()
  normal-mode(t)
  after-find-file(nil t)
  find-file-noselect-1(#<buffer a.cpp> "/tmp/test-ccls/a.cpp" nil nil "/tmp/test-ccls/a.cpp" (140 46))
  find-file-noselect("/tmp/test-ccls/a.cpp" nil nil nil)
  find-file("/tmp/test-ccls/a.cpp")
  dired-find-file()
  funcall-interactively(dired-find-file)
  call-interactively(dired-find-file nil nil)
  command-execute(dired-find-file)
```

The second time:
```
Debugger entered--entering a function:
* lsp()
  (progn (print ccls-initialization-options) (lsp))
  (if (derived-mode-p (quote prog-mode)) (progn (print ccls-initialization-options) (lsp)))
  (closure (t) nil (if (derived-mode-p (quote prog-mode)) (progn (print ccls-initialization-options) (lsp))))()
  run-hooks(hack-local-variables-hook)
  hack-local-variables-apply()
  hack-local-variables(no-mode)
  run-mode-hooks(c++-mode-hook)
  c++-mode()
  set-auto-mode-0(c++-mode nil)
  set-auto-mode()
  normal-mode(t)
  after-find-file(nil t)
  find-file-noselect-1(#<buffer a.cpp> "/tmp/test-ccls/a.cpp" nil nil "/tmp/test-ccls/a.cpp" (140 46))
  find-file-noselect("/tmp/test-ccls/a.cpp" nil nil nil)
  find-file("/tmp/test-ccls/a.cpp")
  dired-find-file()
  funcall-interactively(dired-find-file)
  call-interactively(dired-find-file nil nil)
  command-execute(dired-find-file)
```

Another proof:
```
Command "ccls --log-file=/tmp/ccls.log" is present on the path.
Found the following clients for /tmp/test-ccls/a.cpp: (server-id ccls, priority 0)
The following clients were selected based on priority: (server-id ccls, priority 0)
Command "ccls --log-file=/tmp/ccls2.log" is present on the path.
Found the following clients for /tmp/test-ccls/a.cpp: (server-id ccls, priority 0)
The following clients were selected based on priority: (server-id ccls, priority 0)
Creating watch for /tmp/test-ccls/
```

It seems to be a `spacemacs` issue. I found this hook in spacemacs's `c-c++` layer (packages.el)

I've also found syl20bnr says the `lsp` layer is still under working. So I could only wait for the work finished.

https://github.com/syl20bnr/spacemacs/issues/12784#issuecomment-541327483

**Comment by hg-zt** [comment]
more specifically:

```
(defun c-c++/init-ccls ()
  (use-package ccls
    :if (eq c-c++-backend 'lsp-ccls)
    :config
    (spacemacs//c-c++-lsp-config)
    :hook ((c-mode c++-mode) .
            (lambda () (cl-pushnew #'company-lsp company-backends) (require 'ccls) (remhash 'clangd lsp-clients) (lsp)))))
```

https://github.com/syl20bnr/spacemacs/blob/develop/layers/%2Blang/c-c%2B%2B/packages.el#L292-L298

Does this breaks the intention of `hack-local-variables-hook` ? It called `lsp` before `hack-local-variables-hook`

**Comment by hg-zt** [comment]
Well, at least the latest commit of spacemacs on the `develop` branch does not work for me. 

Just removed `c-c++` layer and use `use-package` instead solves the problem for me.

```
(defun dotspacemacs/user-config ()

  (use-package lsp-mode)
  (use-package lsp-ui :commands lsp-ui-mode)
  (use-package company-lsp :commands company-lsp)
  (use-package ccls)

  (add-hook 'hack-local-variables-hook
            (lambda () (when (derived-mode-p 'prog-mode) (require 'ccls) (lsp))))
  )
)
```

**Comment by Lenbok** [comment]
> Will be solved by #544. It will allow the following configuration:
> 
> ```
> (add-hook 'hack-local-variables-hook (lambda () (when (derived-mode-p 'prog-mode) (lsp))))
> ```

@yyoncho This will try to turn on lsp for all programming modes, if lsp has already been enabled (according to the user preferences for the mode) before the dir locals are read, then perhaps it's more specific to do:
```
(add-hook 'hack-local-variables-hook (lambda () (when lsp-mode (lsp))))
```
